packages/scoped-elements/README.md

```diff
@@ -0,0 +1,261 @@
+
+## How it works
+
+`createScopedHtml` defines your custom element classes and creates a unique scoped tag for each of them if you need to use multiple versions of an element. In case those classes were already defined, it will return the previously defined tag.
+
+Then, the `html` function provided by the `createScopedHtml` method transforms the template literal into another one replacing the tags used by the user by the ones defined by the custom elements. Finally, the transformed template literal is going to be processed by `lit-html`.
+
+`<my-button>${this.text}</my-button>`
+
+becomes:
+`<my-button-3>${this.text}</my-button-3>`
+
+## Usage
+
+1. Import `createScopedHtml` from `@open-wc/scoped-elements`.
+
+   ```js
+   import { createScopedHtml } from '@open-wc/scoped-elements';
+   ```
+
+2. Import the classes of the components you want to use.
+
+   ```js
+   import MyButton from './MyButton.js';
+   import MyPanel from './MyPanel.js';
+   ```
+
+3. Create the `html` function and define the tags you want to use for your
+   components.
+
+   ```js
+   const html = createScopedHtml({
+     'my-button': MyButton,
+     'my-panel': MyPanel,
+   });
+   ```
+
+4. Use your components in your html.
+
+   ```js
+   render() {
+     return html`
+       <my-panel class="panel">
+         <my-button>${this.text}</my-button>
+       </my-panel}>
+     `;
+   }
+   ```
+
+### Complete example
+
+```js
+import { css, LitElement } from 'lit-element';
+import { createScopedHtml } from '@open-wc/scoped-elements';
+import MyButton from './MyButton.js';
+import MyPanel from './MyPanel.js';
+
+const html = createScopedHtml({
+  'my-button': MyButton, // <-- binds the element to a tag in your html
+  'my-panel': MyPanel,
+});
+
+export default class MyElement extends LitElement {
+  static get styles() {
+    return css`
+      .panel {
+        padding: 10px;
+        background-color: grey;
+      }
+    `;
+  }
+
+  static get properties() {
+    return {
+      text: String,
+    };
+  }
+
+  render() {
+    return html`
+      <my-panel class="panel">
+        <my-button>${this.text}</my-button>
+      </my-panel}>
+    `;
+  }
+}
+```
+
+## Limitations
+
+1. Components imported via npm **MUST NOT** be self registering components. If a shared component (installed from npm) does not offer an export to the class alone, without the registration side effect, then this component can not be used. E.g. every component that calls `customElement.define`
+
+   ```js
+   export class MyEl { ... }
+   customElement.define('my-el', MyEl);
+   ```
+
+   Or uses the `customElement` typescript decorator
+
+   ```ts
+   @customElement('my-el')
+   export class MyEl { ... }
+   ```
+
+   Only side effects free class exports may be used
+
+   ```js
+   export class MyEl { ... }
+   ```
+
+   Note: not share or components within the main app shell are exempted from this rule.
```

Comment:
```suggestion
   Note: This rule does not apply to components that you write, and are in control of yourself and do not come from a third party library.
```

Reply:
As I said in my previous comment this note refers to how you must share your component.

I think a good practice is export the component class and define it in separate files, like in Lion library, so when you are going to use the component you can decide to use the self-registered one or register it using scoped-elements feature.

Reply:
I think is better to remove this paragraph by the moment to avoid confussion, don't you think?

Reply:
as i pointed out if we add a check if the returned class from customElements get is === KlassToRegister we can simply ignore all that .

Reply:
this was about that in your app shell (or your main app) you do not need to care about using scoped-elements... e.g. self registering elements are fine in that case... as these component are always unique to the app and can never become double-loaded...

but yeah maybe that is too much in depth... maybe it needs a full app example to get this concept clear 🙈